src/sentry/static/sentry/app/stores/guideStore.jsx

```diff
@@ -210,38 +262,20 @@ const GuideStore = Reflux.createStore({
     }
 
     // Pick the first guide that satisfies conditions.
-    let bestGuideKey = null;
-    let user = ConfigStore.get('user');
-    for (let key of guideKeys) {
-      if (key === ALERT_REMINDER_1) {
-        if (!this.checkAlertTipData()) {
```

Comment:
I believe `this.checkAlertTipData` had some concurrency issues because it *sometimes* had async callbacks to `this.updateCurrentGuide`.

If you look at the `should render tip` test that was failing. I think here's what the callstack could look like with the 3 guides (`Guide1`, `Guide2`, `alert_reminder_1`:

```
onSetActiveProject ->
  updateCurrentGuide ->
    checkAlertTipData (Guide1) ->
    checkAlertTipData (Guide2) ->
    checkAlertTipData (alert_reminder_1) ->
      // stats and rules are undefined here
      **fetch stats** -> ...
      **fetch rules** -> ...
    **store triggers with state**

...
**fetch stats success** ->
  updateCurrentGuide ->
    checkAlertTipData (Guide1) ->
    checkAlertTipData (Guide2) ->
    checkAlertTipData (alert_reminder_1) ->
	  // rules are undefined here
      **fetch rules** -> ...
    **store triggers with state**

**fetch rules success** ->
  updateCurrentGuide ->
    checkAlertTipData (Guide1) ->
    checkAlertTipData (Guide2) ->
    checkAlertTipData (alert_reminder_1) ->
	  // both are defined now
    **store triggers with state**

**fetch rules success** ->
  updateCurrentGuide ->
    checkAlertTipData (Guide1) ->
    checkAlertTipData (Guide2) ->
    checkAlertTipData (alert_reminder_1) ->
	  // both are defined now
    **store triggers with state**
```

I've refactored things a bit so that `updateCurrentGuide` is async and waits until it can fetch project stats and rules before it calls `trigger`. This reduces the number of times a component needs to re-render as well as making it a bit easier to test.

Reply:
looks much cleaner! I tried to repro the test failure locally but couldn't.. what was the issue? and why does triggering rule fetch twice cause problems?

Reply:
It would fail when you run multiple tests (not just `guideStore.spec`)

Reply:
and why does triggering rule fetch twice cause problems?

Reply:
So here's what's happening when things are flattened

```
onSetActiveProject ->
    updateCurrentGuide ->
        checkAlertTipData (alert_reminder_1) ->
            // stats and rules are undefined here
            // `ready` is false
            **fetch stats + success** -> ...
                updateCurrentGuide ->
                    checkAlertTipData (alert_reminder_1) ->
                    // rules are undefined here
                        **fetch rules + success** -> ...
                            updateCurrentGuide ->
                                checkAlertTipData (alert_reminder_1) ->
                                    // both are defined now
                                    // `updateCurrentGuide` finishes and currentGuide is set, store triggers with state
            **fetch rules + success** -> ...
                updateCurrentGuide ->
                    checkAlertTipData (alert_reminder_1) ->
            
            // finally returns with ready = false
        // bestGuideKey is null
        // state.currentGuide = null
```

Reply:
I believe this only happens during testing because jest tries to make things synchronous. In real world, the fetch stats + rules would be pushed to the next execution stack so it'll happen after the first `updateCurrentGuide` gets returned.

Reply:
ooh ya I remember thinking about this and forgetting that jest makes things sync. damn

Reply:
good catch, thanks!

Reply:
is there a way to make jest not do that / would it be better overall if we did that?